File: java/quantile-estimator/src/main/java/org/mydotey/quantile/facade/QuantileEstimators.java
package org.mydotey.quantile.facade;

import java.util.Comparator;

import org.mydotey.quantile.QuantileEstimator;
import org.mydotey.quantile.ckms.CkmsQuantileEstimator;
import org.mydotey.quantile.ckms.CkmsQuantileEstimatorConfig;
import org.mydotey.quantile.classic.ClassicQuantileEstimator;
import org.mydotey.quantile.classic.ClassicQuantileEstimatorConfig;
import org.mydotey.quantile.decorator.ConcurrencyDecorator;
import org.mydotey.quantile.decorator.ValidationDecorator;
import org.mydotey.quantile.gk.GkQuantileEstimator;
import org.mydotey.quantile.gk.GkQuantileEstimatorConfig;
import org.mydotey.quantile.kll.KllQuantileEstimator;
import org.mydotey.quantile.kll.KllQuantileEstimatorConfig;
import org.mydotey.quantile.value.Calculator;

/**
 * @author koqizhao
 *
 * Apr 1, 2018
 */
public class QuantileEstimators {

    protected QuantileEstimators() {

    }

    public static <T> QuantileEstimator<T> newClassicEstimator(Comparator<T> comparator, Calculator<T> calculator) {
        QuantileEstimator<T> quantileEstimator = new ClassicQuantileEstimator<>(
                new ClassicQuantileEstimatorConfig<>(comparator, calculator));
        return decorate(quantileEstimator);
    }

    /**
     * @param k    first compactor size
     * c default to 2.0 / 3.0
     */
    public static <T> QuantileEstimator<T> newKllEstimator(Comparator<T> comparator, int k) {
        QuantileEstimator<T> quantileEstimator = new KllQuantileEstimator<>(
                new KllQuantileEstimatorConfig<>(comparator, k));
        return decorate(quantileEstimator);
    }

    /**
     * @param k    first compactor size
     * @param c    compaction rate
     */
    public static <T> QuantileEstimator<T> newKllEstimator(Comparator<T> comparator, int k, double c) {
        QuantileEstimator<T> quantileEstimator = new KllQuantileEstimator<>(
                new KllQuantileEstimatorConfig<>(comparator, k, c));
        return decorate(quantileEstimator);
    }

    public static <T> CkmsQuantileEstimatorConfig.Builder<T> newCkmsEstimatorConfigBuilder() {
        return CkmsQuantileEstimatorConfig.<T> newBuiler();
    }

    public static <T> QuantileEstimator<T> newCKmsEstimator(CkmsQuantileEstimatorConfig<T> config) {
        QuantileEstimator<T> quantileEstimator = new CkmsQuantileEstimator<>(config);
        return decorate(quantileEstimator);
    }

    public static <T> QuantileEstimator<T> newGkEstimator(Comparator<T> comparator, double error,
            int compactThreshold) {
        QuantileEstimator<T> quantileEstimator = new GkQuantileEstimator<>(
                new GkQuantileEstimatorConfig<>(comparator, error, compactThreshold));
        return decorate(quantileEstimator);
    }

    protected static <T> QuantileEstimator<T> decorate(QuantileEstimator<T> quantileEstimator) {
        ConcurrencyDecorator<T> concurrencyDecorator = new ConcurrencyDecorator<>(quantileEstimator);
        ValidationDecorator<T> validationDecorator = new ValidationDecorator<>(concurrencyDecorator);
        return validationDecorator;
    }

}

File: dotnet/src/MyDotey.QuantileEstimator/Facade/QuantileEstimators.cs
using System;
using System.Collections.Generic;

using MyDotey.Quantile.Value;
using MyDotey.Quantile.Validation;
using MyDotey.Quantile.Concurrent;
using MyDotey.Quantile.TimeWindow;

using MyDotey.Quantile.Classic;
using MyDotey.Quantile.Ckms;
using MyDotey.Quantile.Gk;
using MyDotey.Quantile.Kll;

/**
 * @author koqizhao
 *
 * Apr 1, 2018
 */
namespace MyDotey.Quantile.Facade
{
    public class QuantileEstimators
    {
        protected QuantileEstimators()
        {

        }

        public static IQuantileEstimator<T> NewClassicEstimator<T>(IComparer<T> comparer, ICalculator<T> calculator)
        {
            IQuantileEstimator<T> quantileEstimator = new ClassicQuantileEstimator<T>(
                    new ClassicQuantileEstimatorConfig<T>(comparer, calculator));
            return new ValidationDecorator<T>(quantileEstimator);
        }

        /**
         * @param k    first compactor size
         * c default to 2.0 / 3.0
         */
        public static IQuantileEstimator<T> NewKllEstimator<T>(IComparer<T> comparer, int k)
        {
            IQuantileEstimator<T> quantileEstimator = new KllQuantileEstimator<T>(
                    new KllQuantileEstimatorConfig<T>(comparer, k));
            return new ValidationDecorator<T>(quantileEstimator);
        }

        /**
         * @param k    first compactor size
         * @param c    compaction rate
         */
        public static IQuantileEstimator<T> NewKllEstimator<T>(IComparer<T> comparer, int k, double c)
        {
            IQuantileEstimator<T> quantileEstimator = new KllQuantileEstimator<T>(
                    new KllQuantileEstimatorConfig<T>(comparer, k, c));
            return new ValidationDecorator<T>(quantileEstimator);
        }

        public static CkmsQuantileEstimatorConfig<T>.Builder NewCkmsEstimatorConfigBuilder<T>()
        {
            return CkmsQuantileEstimatorConfig<T>.NewBuilder();
        }

        public static IQuantileEstimator<T> NewCkmsEstimator<T>(CkmsQuantileEstimatorConfig<T> config)
        {
            IQuantileEstimator<T> quantileEstimator = new CkmsQuantileEstimator<T>(config);
            return new ValidationDecorator<T>(quantileEstimator);
        }

        public static IQuantileEstimator<T> NewGkEstimator<T>(IComparer<T> comparer, double error,
                int compactThreshold)
        {
            IQuantileEstimator<T> quantileEstimator = new GkQuantileEstimator<T>(
                    new GkQuantileEstimatorConfig<T>(comparer, error, compactThreshold));
            return new ValidationDecorator<T>(quantileEstimator);
        }

        public static IQuantileEstimator<T> NewTimeWindowEstimator<T>(
                Func<IQuantileEstimator<T>> quantileEstimatorSupplier, long timeWindowMillis,
                long rotateDurationMillis)
        {
            return new TimeWindowQuantileEstimator<T>(new TimeWindowQuantileEstimatorConfig<T>(quantileEstimatorSupplier,
                    timeWindowMillis, rotateDurationMillis));
        }

        public static IQuantileEstimator<T> NewConcurrentEstimator<T>(IQuantileEstimator<T> quantileEstimator)
        {
            return new ConcurrencyDecorator<T>(quantileEstimator);
        }
    }
}